src/SFML/Graphics/VertexBuffer.cpp

```diff
@@ -0,0 +1,373 @@
+m_primitiveType(Points),
+m_usage        (usage)
+{
+}
+
+
+////////////////////////////////////////////////////////////
+VertexBuffer::VertexBuffer(PrimitiveType type, VertexBuffer::Usage usage) :
+m_buffer       (0),
+m_size         (0),
+m_primitiveType(type),
+m_usage        (usage)
+{
+}
+
+
+////////////////////////////////////////////////////////////
+VertexBuffer::VertexBuffer(const VertexBuffer& copy) :
+m_buffer       (0),
+m_size         (0),
+m_primitiveType(copy.m_primitiveType)
+{
+    if (copy.m_buffer && copy.m_size)
+    {
+        if (!create(copy.m_size))
+        {
+            err() << "Could not create vertex buffer for copying" << std::endl;
+            return;
+        }
+
+        if (!update(copy))
+            err() << "Could not copy vertex buffer" << std::endl;
+    }
+}
+
+
+////////////////////////////////////////////////////////////
+VertexBuffer::~VertexBuffer()
+{
+    if (m_buffer)
+    {
+        TransientContextLock contextLock;
+
+        glCheck(GLEXT_glDeleteBuffers(1, &m_buffer));
+    }
+}
+
+
+////////////////////////////////////////////////////////////
+bool VertexBuffer::create(std::size_t vertexCount)
+{
+    if (!isAvailable())
+        return false;
+
+    TransientContextLock contextLock;
+
+    if (!m_buffer)
+        glCheck(GLEXT_glGenBuffers(1, &m_buffer));
+
+    if (!m_buffer)
+    {
+        err() << "Could not create vertex buffer, generation failed" << std::endl;
+        return false;
+    }
+
+    glCheck(GLEXT_glBindBuffer(GLEXT_GL_ARRAY_BUFFER, m_buffer));
+    glCheck(GLEXT_glBufferData(GLEXT_GL_ARRAY_BUFFER, sizeof(Vertex) * vertexCount, 0, usageToGlEnum(m_usage)));
+    glCheck(GLEXT_glBindBuffer(GLEXT_GL_ARRAY_BUFFER, 0));
+
+    m_size = vertexCount;
+
+    return true;
+}
+
+
+////////////////////////////////////////////////////////////
+std::size_t VertexBuffer::getVertexCount() const
+{
+    return m_size;
+}
+
+
+////////////////////////////////////////////////////////////
+bool VertexBuffer::update(const Vertex* vertices)
+{
+    return update(vertices, m_size, 0);
+}
+
+
+////////////////////////////////////////////////////////////
+bool VertexBuffer::update(const Vertex* vertices, std::size_t vertexCount, unsigned int offset)
+{
+    // Sanity checks
+    if (!m_buffer)
+        return false;
+
+    if (!vertices)
+        return false;
+
+    if (offset && (offset + vertexCount > m_size))
+        return false;
+
+    TransientContextLock contextLock;
+
+    glCheck(GLEXT_glBindBuffer(GLEXT_GL_ARRAY_BUFFER, m_buffer));
+
+    // Check if we need to resize or orphan the buffer
+    if (vertexCount >= m_size)
+    {
+        glCheck(GLEXT_glBufferData(GLEXT_GL_ARRAY_BUFFER, sizeof(Vertex) * vertexCount, 0, usageToGlEnum(m_usage)));
+
+        m_size = vertexCount;
+    }
+
+    glCheck(GLEXT_glBufferSubData(GLEXT_GL_ARRAY_BUFFER, sizeof(Vertex) * offset, sizeof(Vertex) * vertexCount, vertices));
+
+    glCheck(GLEXT_glBindBuffer(GLEXT_GL_ARRAY_BUFFER, 0));
```

Comment:
I wonder if we should do like in `sf::Texture` and preserve the previous VBO binding. I don't like this strategy, but at least it avoids surprises for users that mix SFML classes and OpenGL calls.

Reply:
Mixing SFML and OpenGL is already complicated enough as it is. Until now we've told people that we don't mess with buffer binding state and they should make sure not to touch it or at least unbind anything that they bound. Changing it again would just make them more confused than they already were before.

I've already been thinking of a way for users to be able to draw in a separate context but to the same window framebuffer as the sf::RenderTarget. This way both SFML and user OpenGL can live in harmony without ever having to know about what the other is doing. This will deprecate the pushGLStates and popGLStates that is still necessary now.

Reply:
> I've already been thinking of a way for users to be able to draw in a separate context but to the same window framebuffer as the sf::RenderTarget

That would be the best option. Is it feasible?

Reply:
Yes... but not in this PR. 😄